ccl/changefeedccl: TestChangefeedTruncateOrDrop failed

ccl/changefeedccl.TestChangefeedTruncateOrDrop [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3601373&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3601373&tab=artifacts#/) on master @ [649a8ca38cdc0883dd7a7811b189ea53af19f644](https://github.com/cockroachdb/cockroach/commits/649a8ca38cdc0883dd7a7811b189ea53af19f644):


```
=== RUN   TestChangefeedTruncateOrDrop
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestChangefeedTruncateOrDrop801199702
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestChangefeedTruncateOrDrop
    changefeed_test.go:2122: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestChangefeedTruncateOrDrop801199702
--- FAIL: TestChangefeedTruncateOrDrop (10.84s)
=== RUN   TestChangefeedTruncateOrDrop/sinkless
    changefeed_test.go:2113: expected ""drop" was dropped" error got: <nil>
    --- FAIL: TestChangefeedTruncateOrDrop/sinkless (2.03s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestChangefeedTruncateOrDrop PKG=./pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/cdc HonoreDB
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedTruncateOrDrop.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -2063,6 +2063,13 @@ func TestChangefeedTruncateOrDrop(t *testing.T) {
 		registry := f.Server().JobRegistry().(*jobs.Registry)
 		metrics := registry.MetricsStruct().Changefeed.(*Metrics)
 
+		drainUntilErr := func(f cdctest.TestFeed) (err error) {
+			var msg *cdctest.TestFeedMessage
+			for msg, err = f.Next(); msg != nil; {
+			}
+			return err
+		}
+
 		sqlDB.Exec(t, `CREATE TABLE truncate (a INT PRIMARY KEY)`)
 		sqlDB.Exec(t, `CREATE TABLE truncate_cascade (b INT PRIMARY KEY REFERENCES truncate (a))`)
 		sqlDB.Exec(t,
@@ -2074,10 +2081,10 @@ func TestChangefeedTruncateOrDrop(t *testing.T) {
 		assertPayloads(t, truncate, []string{`truncate: [1]->{"after": {"a": 1}}`})
 		assertPayloads(t, truncateCascade, []string{`truncate_cascade: [1]->{"after": {"b": 1}}`})
 		sqlDB.Exec(t, `TRUNCATE TABLE truncate CASCADE`)
-		if _, err := truncate.Next(); !testutils.IsError(err, `"truncate" was truncated`) {
+		if err := drainUntilErr(truncate); !testutils.IsError(err, `"truncate" was truncated`) {
 			t.Fatalf(`expected ""truncate" was truncated" error got: %+v`, err)
 		}
-		if _, err := truncateCascade.Next(); !testutils.IsError(
+		if err := drainUntilErr(truncateCascade); !testutils.IsError(
 			err, `"truncate_cascade" was truncated`,
 		) {
 			t.Fatalf(`expected ""truncate_cascade" was truncated" error got: %+v`, err)
@@ -2090,7 +2097,7 @@ func TestChangefeedTruncateOrDrop(t *testing.T) {
 		defer closeFeed(t, drop)
 		assertPayloads(t, drop, []string{`drop: [1]->{"after": {"a": 1}}`})
 		sqlDB.Exec(t, `DROP TABLE drop`)
-		if _, err := drop.Next(); !testutils.IsError(err, `"drop" was dropped`) {
+		if err := drainUntilErr(drop); !testutils.IsError(err, `"drop" was dropped`) {
 			t.Errorf(`expected ""drop" was dropped" error got: %+v`, err)
 		}
 		assertFailuresCounter(t, metrics, 3)

